### DISCUSSION TITLE
How to cycle with vim copilot?

### DISCUSSION THREAD

**Discussion by dino-rodriguez** [community/community#44941] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

I'm currently using copilot in vim with: https://github.com/github/copilot.vim

My understanding is that to cycle, you need to use ALT-]

When I try to do this in vim (on Mac), it does not work. I've also tried turning my alt key on meta/esc+ and it doesn't seem to make a difference

Has anybody gotten this to work on Mac, or found a way to remap this effectively?

**Comment by jge162** [comment]
hi @dino-rodriguez On a Mac, the Option key is often used as a replacement for the Alt key. You can try remapping the keybinding to use the Option key instead by adding the following line to your .vimrc file:

```nmap <M-]> <D-]>```

just remember to restart your vim after making the changes to your .vimrc file.

**Reply by jge162** [reply]
also default keybinding for cycling through suggestions in copilot.vim is <M-]>

**Reply by dino-rodriguez** [reply]
I tried adding this keybinding to vim: `nmap <M-]> <D-]>`

On normal mode in iterm, it just types the special character ‘

On esc+ mode, when I type option ] it just removes the current suggestion instead of cycling through

**Reply by dino-rodriguez** [reply]
Also what is this default keybinding: <M-]> ?

Is that not just option + ] ?

**Reply by jge162** [reply]
Default is M-J

**Reply by dino-rodriguez** [reply]
Sorry not sure I'm following, if I type M and J it just types those letters

**Reply by jge162** [reply]
> Sorry not sure I'm following, if I type M and J it just types those letters

sorry typo above. 
@dino-rodriguez The keybinding you are trying to add is mapping the Meta key (M) and the close square bracket key (]) to the sequence of the Esc key and the close square bracket key (D-]).

It seems that the issue is with the terminal (iTerm) not recognizing the Meta key. You can try using the 'Ctrl' key instead of 'Meta' in your keybinding.

Also, the 'D-' is not a valid key sequence it is probably not working as expected. You can try using '<C-]>' instead of '<D-]>'

**Reply by dino-rodriguez** [reply]
So I set the keybinding to this and am trying it. Seems to just delete my suggestion when I press ctrl and [
![Kapture 2023-01-21 at 16 34 23](https://user-images.githubusercontent.com/10410068/213887949-37376430-9cae-46f9-8583-e7e8a4dea70f.gif)

**Reply by dino-rodriguez** [reply]
Not sure if this is a bug in copilot or still something i am doing on my end

**Reply by jge162** [reply]
Hm man this seems rough. It is possible that the problem is that the keybinding is conflicting with a default keybinding in Vim. For example, the keybinding <C-[> is used to enter the Normal mode from Insert mode. It could also be that the terminal is not recognizing the keybinding.

You can try using a different keybinding, such as "nmap <C-j> <C-]>", which maps the key combination of "Ctrl" and "j" to the key combination of "Ctrl" and "]".

**Reply by dino-rodriguez** [reply]
Hm. I tried that keybinding and it also doesn't seem to work. Dang - not sure what's going on here.

Thanks for the help though. I appreciate it!

**Reply by jge162** [reply]
Sorry we couldn’t figure it out.

**Reply by dino-rodriguez** [reply]
Figured it out! See the answer I posted below

**Comment by dino-rodriguez** [comment, marked as answer]
Figured it out. Not sure why the default keybindings aren't working in esc+ mode, but I switched back to normal mode and realized I was accidentally using `nnoremap` instead of `inoremap` (which is for insert mode) & also directly mapped the copilot commands

```
"copilot keybindings
"ALT-]
inoremap ‘ <Plug>(copilot-next)
"ALT-[
inoremap “ <Plug>(copilot-previous)
"ALT-\
inoremap « <Plug>(copilot-suggest)
```

Explicitly remapping the copilot commands worked.

**Reply by jge162** [reply]
Sweet

**Comment by CervEdin** [comment]
Why are these the default key mappings?
